GSoC: Improved GitHub Checks project page (#3226)
jenkins-infra/jenkins.io: content/projects/gsoc/2020/projects/github-checks.adoc

  idea: /projects/gsoc/2020/project-ideas/github-checks
---

> The link:https://developer.github.com/v3/checks/[GitHub Checks API] allows developers to report the CI integrations’ detail information rather than the binary pass/fail build status on GitHub pages. This project is about implementing this API as a new Jenkins plugin. By consuming this API, other plugins can easily create GitHub checks. Thus, any information during the Jenkins CI process like warnings, summaries, and durations can be directly shown on GitHub pages.

Reviewer: Best practice: add line breaks so that each sentence starts on a new line

Reviewer: Also keep in mind that "Jenkins CI" is a deprecated expression. Better to just use "Jenkins"

=== Project Deliverables
* A new Jenkins plugin that implements GitHub Checks API 

> * Pipeline supprt feature for GitHub Checks API

Reviewer: ```suggestion
* Pipeline support feature for GitHub Checks API
```

> * A new feature for Warnings Next Generation plugin to report warnings to GitHub by consuming the API

Reviewer: ```suggestion
* A new feature for plugin:warnings-ng[Warnings Next Generation plugin] to report warnings to GitHub by consuming the API
```

It would be nice to hyperlink components to simplify navigation. This macro will create a link to https://plugins.jenkins.io/warnings-ng/

Author: These predefined macros are cool!

* A new feature for Code Coverage API plugin to report code coverage to GitHub by consuming the API

> * A general API defined for similar concepts in other platforms (GitHub, GitLab, BitBucket) if possible

Reviewer: I would highlight "if possible" here:

```suggestion
* Maybe: A general API defined for similar concepts in other platforms (GitHub, GitLab, BitBucket)
```

> === Phase1 Schedule

Reviewer: Do not hesitate to use additional heading levels to group things :)

* Implement the Checks API with pipeline support in a new Jenkins plugin.
* Blog post and presentation slides for phase 1.

=== Phase2 Schedule
* Consume the API in Warnings Next Generation plugin as a new feature.
* Consume the API in Code Coverage API plugin as a new feature.
* Blog post and presentation slides for phase 2.

=== Phase3 Schedule
* Add pipeline support for this GitHub Checks API plugin.
* Define general API for other platforms (GitHub, GitLab, BitBucket).
* Consuming and implementation guide.
* Final blog post and presentation slides.

=== Architecture
* The diagram below shows the class architecture:

image:/images/post-images/gsoc-github-checks/architecture.png[title="GitHub Checks API Plugin Architecture" role="center" width=1000,height=500]

=== Workflow
* These properties are needed when creating a specific GitHub Checks:

>   ** APP indentifire: Used to authenticate as a GitHub app in order to request access tokens for an installation (a user/organization account that installs the app on repositories) of  the app. This property is created by GitHub when registering Jenkins as a new GitHub App;

Reviewer: ```suggestion
  ** APP identifier: Used to authenticate as a GitHub app in order to request access tokens for an installation (a user/organization account that installs the app on repositories) of  the app. This property is created by GitHub when registering Jenkins as a new GitHub App;
```

>   ** App installation id: Used to authenticate as an installation. We can easily get this property from link:https://developer.github.com/v3/activity/events/types/#installationrepositoriesevent[Installtion Repository Event] when the users install the app on their GitHub repositories;

Reviewer 2: ```suggestion
  ** App installation id: Used to authenticate as an installation. We can easily get this property from link:https://developer.github.com/v3/activity/events/types/#installationrepositoriesevent[Installation Repository Event] when the users install the app on their GitHub repositories;
```
